File: packages/numwasm/benchmark/python/linalg/operations.py
"""
Linear algebra benchmarks: dot, matmul, norm, det, inv
"""

import numpy as np
from ..lib.benchmark_runner import BenchmarkRunner
from ..lib.config import BENCHMARK_CONFIG, MATRIX_SIZES, RANDOM_SEED


def benchmark_dot(runner: BenchmarkRunner) -> dict:
    """Benchmark np.dot for vectors."""
    results = []

    for size in MATRIX_SIZES:
        print(f"    dot size={size}... ", end='', flush=True)

        np.random.seed(RANDOM_SEED)
        arr1 = np.random.randn(size)
        np.random.seed(RANDOM_SEED + 1)
        arr2 = np.random.randn(size)

        result = runner.run(lambda: np.dot(arr1, arr2))

        results.append({
            'params': {'size': size},
            'meanMs': result.mean_ms,
            'minMs': result.min_ms,
            'maxMs': result.max_ms,
            'stdMs': result.std_ms,
            'iterations': result.iterations,
            'result': result.result,
        })

        print(f"{result.mean_ms:.4f} ms ({result.iterations} iters)")

    return {'id': 'dot', 'name': 'Dot Product', 'results': results}


def benchmark_matmul(runner: BenchmarkRunner) -> dict:
    """Benchmark np.matmul for matrices."""
    results = []

    for size in MATRIX_SIZES:
        print(f"    matmul size={size}x{size}... ", end='', flush=True)

        np.random.seed(RANDOM_SEED)
        mat1 = np.random.randn(size, size)
        np.random.seed(RANDOM_SEED + 1)
        mat2 = np.random.randn(size, size)

        result = runner.run(lambda: np.matmul(mat1, mat2))

        results.append({
            'params': {'size': size},
            'meanMs': result.mean_ms,
            'minMs': result.min_ms,
            'maxMs': result.max_ms,
            'stdMs': result.std_ms,
            'iterations': result.iterations,
        })

        print(f"{result.mean_ms:.4f} ms ({result.iterations} iters)")

    return {'id': 'matmul', 'name': 'Matrix Multiply', 'results': results}


def benchmark_norm(runner: BenchmarkRunner) -> dict:
    """Benchmark np.linalg.norm for matrices."""
    results = []

    for size in MATRIX_SIZES:
        print(f"    norm size={size}x{size}... ", end='', flush=True)

        np.random.seed(RANDOM_SEED)
        mat = np.random.randn(size, size)

        result = runner.run(lambda: np.linalg.norm(mat))

        results.append({
            'params': {'size': size},
            'meanMs': result.mean_ms,
            'minMs': result.min_ms,
            'maxMs': result.max_ms,
            'stdMs': result.std_ms,
            'iterations': result.iterations,
            'result': result.result,
        })

        print(f"{result.mean_ms:.4f} ms ({result.iterations} iters)")

    return {'id': 'norm', 'name': 'Matrix Norm', 'results': results}


def benchmark_det(runner: BenchmarkRunner) -> dict:
    """Benchmark np.linalg.det for matrices."""
    results = []
    # Use smaller sizes for det (O(n^3) operation)
    det_sizes = [s for s in MATRIX_SIZES if s <= 200]

    for size in det_sizes:
        print(f"    det size={size}x{size}... ", end='', flush=True)

        np.random.seed(RANDOM_SEED)
        mat = np.random.randn(size, size)

        result = runner.run(lambda: np.linalg.det(mat))

        results.append({
            'params': {'size': size},
            'meanMs': result.mean_ms,
            'minMs': result.min_ms,
            'maxMs': result.max_ms,
            'stdMs': result.std_ms,
            'iterations': result.iterations,
            'result': result.result,
        })

        print(f"{result.mean_ms:.4f} ms ({result.iterations} iters)")

    return {'id': 'det', 'name': 'Determinant', 'results': results}


def benchmark_inv(runner: BenchmarkRunner) -> dict:
    """Benchmark np.linalg.inv for matrices."""
    results = []
    # Use smaller sizes for inv (O(n^3) operation)
    inv_sizes = [s for s in MATRIX_SIZES if s <= 200]

    for size in inv_sizes:
        print(f"    inv size={size}x{size}... ", end='', flush=True)

        np.random.seed(RANDOM_SEED)
        mat = np.random.randn(size, size)
        # Add diagonal dominance for well-conditioned matrix
        mat = mat + np.eye(size) * size

        result = runner.run(lambda: np.linalg.inv(mat))

        results.append({
            'params': {'size': size},
            'meanMs': result.mean_ms,
            'minMs': result.min_ms,
            'maxMs': result.max_ms,
            'stdMs': result.std_ms,
            'iterations': result.iterations,
        })

        print(f"{result.mean_ms:.4f} ms ({result.iterations} iters)")

    return {'id': 'inv', 'name': 'Matrix Inverse', 'results': results}


def run_linalg_benchmarks() -> list[dict]:
    """Run all linear algebra benchmarks."""
    runner = BenchmarkRunner(BENCHMARK_CONFIG)
    results = []

    print("  Benchmarking dot...")
    results.append(benchmark_dot(runner))

    print("  Benchmarking matmul...")
    results.append(benchmark_matmul(runner))

    print("  Benchmarking norm...")
    results.append(benchmark_norm(runner))

    print("  Benchmarking det...")
    results.append(benchmark_det(runner))

    print("  Benchmarking inv...")
    results.append(benchmark_inv(runner))

    return results


if __name__ == '__main__':
    import json
    results = run_linalg_benchmarks()
    print(json.dumps(results, indent=2))

File: packages/numwasm/benchmark/ts/linalg/operations.ts
/**
 * Linear algebra benchmarks: dot, matmul, norm, det, inv
 */

import { BenchmarkRunner } from '../lib/BenchmarkRunner.js';
import { BENCHMARK_CONFIG, MATRIX_SIZES, RANDOM_SEED } from '../lib/config.js';
import { generateRandomData } from '../lib/utils.js';
import type { OperationResults, BenchmarkDataPoint } from '../lib/types.js';

// @ts-expect-error - Types are in dist/index.d.ts but TS can't resolve mjs
import { NDArray, dot, matmul, loadWasmModule } from '../../../numjs.mjs';
// @ts-expect-error
import * as linalg from '../../../numjs.mjs';

async function benchmarkDot(runner: BenchmarkRunner): Promise<OperationResults> {
  const results: BenchmarkDataPoint[] = [];

  for (const size of MATRIX_SIZES) {
    process.stdout.write(`    dot size=${size}... `);

    const data1 = generateRandomData(size, RANDOM_SEED);
    const data2 = generateRandomData(size, RANDOM_SEED + 1);

    const arr1 = await NDArray.fromArray(data1);
    const arr2 = await NDArray.fromArray(data2);

    try {
      const result = await runner.run(
        () => ({ arr1, arr2 }),
        (setup) => dot(setup.arr1, setup.arr2)
      );

      results.push({
        params: { size },
        meanMs: result.meanMs,
        minMs: result.minMs,
        maxMs: result.maxMs,
        stdMs: result.stdMs,
        iterations: result.iterations,
        result: result.result,
      });

      console.log(`${result.meanMs.toFixed(4)} ms (${result.iterations} iters)`);
    } finally {
      arr1.dispose();
      arr2.dispose();
    }
  }

  return { id: 'dot', name: 'Dot Product', results };
}

async function benchmarkMatmul(runner: BenchmarkRunner): Promise<OperationResults> {
  const results: BenchmarkDataPoint[] = [];

  for (const size of MATRIX_SIZES) {
    process.stdout.write(`    matmul size=${size}x${size}... `);

    const data1 = generateRandomData(size * size, RANDOM_SEED);
    const data2 = generateRandomData(size * size, RANDOM_SEED + 1);

    const arr1 = await NDArray.fromArray(data1);
    const arr2 = await NDArray.fromArray(data2);
    const mat1 = arr1.reshape([size, size]);
    const mat2 = arr2.reshape([size, size]);
    arr1.dispose();
    arr2.dispose();

    try {
      const result = await runner.run(
        () => ({ mat1, mat2 }),
        (setup) => {
          const res = matmul(setup.mat1, setup.mat2);
          const val = res.sum();
          res.dispose();
          return val;
        }
      );

      results.push({
        params: { size },
        meanMs: result.meanMs,
        minMs: result.minMs,
        maxMs: result.maxMs,
        stdMs: result.stdMs,
        iterations: result.iterations,
        result: result.result,
      });

      console.log(`${result.meanMs.toFixed(4)} ms (${result.iterations} iters)`);
    } finally {
      mat1.dispose();
      mat2.dispose();
    }
  }

  return { id: 'matmul', name: 'Matrix Multiply', results };
}

async function benchmarkNorm(runner: BenchmarkRunner): Promise<OperationResults> {
  const results: BenchmarkDataPoint[] = [];

  for (const size of MATRIX_SIZES) {
    process.stdout.write(`    norm size=${size}x${size}... `);

    const data = generateRandomData(size * size, RANDOM_SEED);
    const arr = await NDArray.fromArray(data);
    const mat = arr.reshape([size, size]);
    arr.dispose();

    try {
      const result = await runner.run(
        () => mat,
        (m) => linalg.norm(m)
      );

      results.push({
        params: { size },
        meanMs: result.meanMs,
        minMs: result.minMs,
        maxMs: result.maxMs,
        stdMs: result.stdMs,
        iterations: result.iterations,
        result: result.result,
      });

      console.log(`${result.meanMs.toFixed(4)} ms (${result.iterations} iters)`);
    } finally {
      mat.dispose();
    }
  }

  return { id: 'norm', name: 'Matrix Norm', results };
}

async function benchmarkDet(runner: BenchmarkRunner): Promise<OperationResults> {
  const results: BenchmarkDataPoint[] = [];
  // Use smaller sizes for det (O(n^3) operation)
  const detSizes = MATRIX_SIZES.filter(s => s <= 200);

  for (const size of detSizes) {
    process.stdout.write(`    det size=${size}x${size}... `);

    const data = generateRandomData(size * size, RANDOM_SEED);
    const arr = await NDArray.fromArray(data);
    const mat = arr.reshape([size, size]);
    arr.dispose();

    try {
      const result = await runner.run(
        () => mat,
        (m) => linalg.det(m)
      );

      results.push({
        params: { size },
        meanMs: result.meanMs,
        minMs: result.minMs,
        maxMs: result.maxMs,
        stdMs: result.stdMs,
        iterations: result.iterations,
        result: result.result,
      });

      console.log(`${result.meanMs.toFixed(4)} ms (${result.iterations} iters)`);
    } finally {
      mat.dispose();
    }
  }

  return { id: 'det', name: 'Determinant', results };
}

async function benchmarkInv(runner: BenchmarkRunner): Promise<OperationResults> {
  const results: BenchmarkDataPoint[] = [];
  // Use smaller sizes for inv (O(n^3) operation)
  const invSizes = MATRIX_SIZES.filter(s => s <= 200);

  for (const size of invSizes) {
    process.stdout.write(`    inv size=${size}x${size}... `);

    // Create a well-conditioned matrix (add diagonal dominance)
    const data = generateRandomData(size * size, RANDOM_SEED);
    const arr = await NDArray.fromArray(data);
    const mat = arr.reshape([size, size]);
    arr.dispose();

    // Add identity * size to ensure invertibility
    const eye = linalg.eye(size);
    const scaled = linalg.multiply(eye, size);
    const wellConditioned = linalg.add(mat, scaled);
    mat.dispose();
    eye.dispose();
    scaled.dispose();

    try {
      const result = await runner.run(
        () => wellConditioned,
        (m) => {
          const res = linalg.inv(m);
          const val = res.sum();
          res.dispose();
          return val;
        }
      );

      results.push({
        params: { size },
        meanMs: result.meanMs,
        minMs: result.minMs,
        maxMs: result.maxMs,
        stdMs: result.stdMs,
        iterations: result.iterations,
        result: result.result,
      });

      console.log(`${result.meanMs.toFixed(4)} ms (${result.iterations} iters)`);
    } finally {
      wellConditioned.dispose();
    }
  }

  return { id: 'inv', name: 'Matrix Inverse', results };
}

export async function runLinalgBenchmarks(): Promise<OperationResults[]> {
  console.log('  Loading WASM module...');
  await loadWasmModule();

  const runner = new BenchmarkRunner(BENCHMARK_CONFIG);
  const results: OperationResults[] = [];

  console.log('  Benchmarking dot...');
  results.push(await benchmarkDot(runner));

  console.log('  Benchmarking matmul...');
  results.push(await benchmarkMatmul(runner));

  console.log('  Benchmarking norm...');
  results.push(await benchmarkNorm(runner));

  console.log('  Benchmarking det...');
  results.push(await benchmarkDet(runner));

  console.log('  Benchmarking inv...');
  results.push(await benchmarkInv(runner));

  return results;
}
